File: src/mvmctl/core/_shared/_loopmount/_provisioner.py
"""
Loop-mount provisioner — accumulates provisioning operations and executes
them via the mvm-provision binary.

Content generation is delegated to ``_provisioner/_content.py`` (shared with
all backends).  This class only handles serialization and binary execution::

    lp = LoopMountProvisioner("/path/to/rootfs.ext4", "ext4")
    lp.resize(8_589_934_592)
    lp.set_hostname("my-vm")
    lp.inject_dns(dns_server="10.0.0.1")
    lp.setup_ssh("myuser", pubkeys)
    lp.disable_cloud_init()
    lp.run()   # → build ops → serialize JSON → subprocess → binary
"""

from __future__ import annotations

import base64
import logging
from pathlib import Path

from mvmctl.core._shared._loopmount._manager import LoopMountManager
from mvmctl.core._shared._provisioner._content import (
    ChrootOp,
    CopyDirOp,
    FileOp,
    Operation,
    ProvisionerContent,
    ResizeOp,
)

logger = logging.getLogger(__name__)


class LoopMountProvisioner:
    """
    Accumulates provisioning operations and executes via the mvm-provision binary.

    All content generation is delegated to the shared ``_content.py`` builders.
    This class only stores the pre-built ops and serializes them to the JSON
    protocol expected by the loop-mount binary.
    """

    def __init__(self, rootfs_path: Path, fs_type: str) -> None:
        self._rootfs_path = rootfs_path
        self._fs_type = fs_type
        self._ops: list[Operation] = []

    # -- builder methods --------------------------------------------------

    def resize(self, target_size_bytes: int) -> None:
        """Queue a rootfs resize operation."""
        self._ops.extend(ProvisionerContent.build_resize_ops(target_size_bytes))

    def set_hostname(self, hostname: str) -> None:
        """Queue hostname + /etc/hosts setup."""
        self._ops.extend(ProvisionerContent.build_hostname_ops(hostname))

    def inject_dns(self, *, dns_server: str) -> None:
        """Queue DNS resolver injection."""
        self._ops.extend(ProvisionerContent.build_dns_ops(dns_server))

    def setup_ssh(self, user: str, ssh_pubkeys: list[str]) -> None:
        """Queue SSH key, config, and host-key generation."""
        self._ops.extend(ProvisionerContent.build_ssh_ops(user, ssh_pubkeys))

    def disable_cloud_init(self) -> None:
        """Queue cloud-init datasource blocking + service masking."""
        self._ops.extend(ProvisionerContent.build_cloud_init_disable_ops())

    def inject_cloud_init(self, cloud_init_dir: Path) -> None:
        """Queue cloud-init seed directory injection."""
        self._ops.extend(
            ProvisionerContent.build_cloud_init_inject_ops(cloud_init_dir)
        )

    # -- execution ---------------------------------------------------------

    def run(self) -> None:
        """Execute all queued operations by spawning the mvm-provision binary."""
        files: list[dict[str, object]] = []
        commands: list[str] = []
        copy_dirs: list[dict[str, object]] = []
        resize: dict[str, object] | None = None

        for op in self._ops:
            match op:
                case FileOp():
                    files.append(
                        {
                            "path": op.path,
                            "data": base64.b64encode(op.data).decode("ascii"),
                            "mode": op.mode,
                            "uid": op.uid,
                            "gid": op.gid,
                        }
                    )
                case ChrootOp():
                    commands.append(op.command)
                case CopyDirOp():
                    copy_dirs.append({"src": op.src, "dst": op.dst})
                case ResizeOp():
                    # Only one resize operation expected — take the last one
                    resize = {"action": op.action, "bytes": op.bytes}

        LoopMountManager.execute(
            image_path=str(self._rootfs_path),
            fs_type=self._fs_type,
            files=files if files else None,
            commands=commands if commands else None,
            copy_dirs=copy_dirs if copy_dirs else None,
            resize=resize,
        )
        logger.info("Loop-mount provisioning succeeded")

File: internal/infra/loopmount/provisioner.go
package loopmount

import (
	"context"
	"log/slog"
	"os"

	"mvmctl/internal/infra/provisionercontent"

	loopmountsvc "mvmctl/internal/service/loopmount"
)

// LoopMountProvisioner accumulates provisioning operations and executes
// them via the loopmount service layer (internal/service/loopmount).
//
// All content generation is delegated to the shared provisionercontent.ProvisionerContent
// builders. This class only stores the pre-built ops and serializes them to the
// service types expected by the loopmount provisioner.
//
// Matches Python's LoopMountProvisioner in _loopmount/_provisionercontent.py.
type LoopMountProvisioner struct {
	rootfsPath string
	fsType     string
	cacheDir   string
	ops        []provisionercontent.Operation
}

// NewLoopMountProvisioner creates a new LoopMountProvisioner.
// Matches Python's __init__().
func NewLoopMountProvisioner(rootfsPath string, fsType string, cacheDir string) *LoopMountProvisioner {
	return &LoopMountProvisioner{
		rootfsPath: rootfsPath,
		fsType:     fsType,
		cacheDir:   cacheDir,
		ops:        nil,
	}
}

// ---------------------------------------------------------------------------
// Builder methods — queue provisioning operations
// ---------------------------------------------------------------------------

// Resize queues a rootfs resize operation.
// Matches Python's resize().
func (lp *LoopMountProvisioner) Resize(targetSizeBytes int64) {
	pc := provisionercontent.ProvisionerContent{}
	lp.ops = append(lp.ops, pc.BuildResizeOps(targetSizeBytes)...)
}

// SetHostname queues hostname + /etc/hosts setup.
// Matches Python's set_hostname().
func (lp *LoopMountProvisioner) SetHostname(hostname string) {
	pc := provisionercontent.ProvisionerContent{}
	lp.ops = append(lp.ops, pc.BuildHostnameOps(hostname)...)
}

// InjectDNS queues DNS resolver injection.
// Matches Python's inject_dns().
func (lp *LoopMountProvisioner) InjectDNS(dnsServer string) {
	pc := provisionercontent.ProvisionerContent{}
	lp.ops = append(lp.ops, pc.BuildDNSOps(dnsServer)...)
}

// SetupSSH queues SSH key, config, and host-key generation.
// Matches Python's setup_ssh().
func (lp *LoopMountProvisioner) SetupSSH(user string, sshPubkeys []string) {
	pc := provisionercontent.ProvisionerContent{}
	lp.ops = append(lp.ops, pc.BuildSSHOps(user, sshPubkeys)...)
}

// DisableCloudInit is a no-op at VM creation time.
// Cloud-init is disabled at image import time via BuildDeblobOps.
// Matches Python's disable_cloud_init().
func (lp *LoopMountProvisioner) DisableCloudInit() {
	// No-op — cloud-init is disabled during image import, not VM creation
}

// InjectCloudInit queues cloud-init seed directory injection.
// Matches Python's inject_cloud_init().
func (lp *LoopMountProvisioner) InjectCloudInit(cloudInitDir string) {
	pc := provisionercontent.ProvisionerContent{}
	lp.ops = append(lp.ops, pc.BuildCloudInitInjectOps(cloudInitDir)...)
}

// ---------------------------------------------------------------------------
// Filesystem conversion
// ---------------------------------------------------------------------------

// ConvertTo converts the image to a different filesystem type in-place.
// This is an independent action — it bypasses the regular provisioning ops flow
// and calls the loopmount service directly.
//
// Matches Python's convert_to().
func (lp *LoopMountProvisioner) ConvertTo(ctx context.Context, targetFS string) (map[string]any, error) {
	svc := loopmountsvc.NewProvisioner(lp.cacheDir)
	results, err := svc.Execute(ctx, []loopmountsvc.Op{
		{
			Image:    lp.rootfsPath,
			Action:   "convert_fs",
			TargetFS: targetFS,
		},
	})
	if err != nil {
		return nil, err
	}
	if len(results) == 0 {
		return nil, nil
	}
	r := results[0]
	return map[string]any{
		"new_fs_type":    r.NewFSType,
		"new_size_bytes": r.NewSizeBytes,
	}, nil
}

// ---------------------------------------------------------------------------
// Execution
// ---------------------------------------------------------------------------

// Run executes all queued operations via the loopmount service.
// Matches Python's run() exactly — always invokes the provisioner even with empty ops.
func (lp *LoopMountProvisioner) Run(ctx context.Context) error {
	// Convert queued operations to service types
	op := loopmountsvc.Op{
		Image:  lp.rootfsPath,
		FsType: lp.fsType,
	}

	for _, o := range lp.ops {
		switch o := o.(type) {
		case provisionercontent.FileOp:
			mode := o.Mode
			if mode == 0 {
				mode = 0644
			}
			op.Files = append(op.Files, loopmountsvc.FileOp{
				Path: o.Path,
				Data: o.Data,
				Mode: os.FileMode(mode),
				UID:  o.UID,
				GID:  o.GID,
			})
		case provisionercontent.ChrootOp:
			op.Commands = append(op.Commands, o.Command)
		case provisionercontent.CopyDirOp:
			op.CopyDirs = append(op.CopyDirs, loopmountsvc.CopyDirOp{
				Src:  o.Src,
				Dst:  o.Dst,
				Mode: 0755,
			})
		case provisionercontent.ResizeOp:
			op.Resize = &loopmountsvc.ResizeOp{
				Action: string(o.Action),
				Bytes:  o.Bytes,
			}
		}
	}

	svc := loopmountsvc.NewProvisioner(lp.cacheDir)
	_, err := svc.Execute(ctx, []loopmountsvc.Op{op})
	if err != nil {
		// Pass through the error from Execute without wrapping — matches Python's
		// LoopMountProvisioner.run() which lets errors propagate as-is.
		return err
	}

	slog.Info("Loop-mount provisioning succeeded")
	return nil
}

// ---------------------------------------------------------------------------
// Cleanup
// ---------------------------------------------------------------------------

// CleanupMount unmounts and removes a stale provision mount point.
// Matches Python's cleanup_mount() static method.
func CleanupMount(ctx context.Context, cacheDir string, mountPoint string) bool {
	return loopmountsvc.CleanupMount(mountPoint)
}

// QueueOps appends raw operation objects to the internal ops queue.
// Matches Python's _queue_ops() in _backend.py which does self._lp._ops.extend(ops).
func (lp *LoopMountProvisioner) QueueOps(ops []provisionercontent.Operation) {
	lp.ops = append(lp.ops, ops...)
}

// RootfsPath returns the rootfs path for this provisionercontent.
func (lp *LoopMountProvisioner) RootfsPath() string {
	return lp.rootfsPath
}

// FsType returns the filesystem type for this provisionercontent.
func (lp *LoopMountProvisioner) FsType() string {
	return lp.fsType
}
